src/Shared/test/Shared.Tests/AdaptiveCapacityDictionaryTests.cs

```diff
@@ -0,0 +1,1375 @@
+            Assert.Equal(2, dict.Count);
+            Assert.False(dict.ContainsKey("key"));
+            Assert.True(dict.ContainsKey("other"));
+            Assert.True(dict.ContainsKey("dotnet"));
+            Assert.IsType<KeyValuePair<string, object?>[]>(dict._arrayStorage);
+        }
+
+        [Fact]
+        public void Remove_KeyAndOutValue_ListStorage_KeyExists_Last()
+        {
+            // Arrange
+            object value = "value";
+            var dict = new AdaptiveCapacityDictionary<string, object>()
+            {
+                { "other", 5 },
+                { "dotnet", "rocks" },
+                { "key", value }
+            };
+
+            // Act
+            var result = dict.Remove("key", out var removedValue);
+
+            // Assert
+            Assert.True(result);
+            Assert.Same(value, removedValue);
+            Assert.Equal(2, dict.Count);
+            Assert.False(dict.ContainsKey("key"));
+            Assert.True(dict.ContainsKey("other"));
+            Assert.True(dict.ContainsKey("dotnet"));
+            Assert.IsType<KeyValuePair<string, object?>[]>(dict._arrayStorage);
+        }
+
+        [Fact]
+        public void TryAdd_EmptyStringIsAllowed()
+        {
+            // Arrange
+            var dict = new AdaptiveCapacityDictionary<string, string>();
+
+            // Act
+            var result = dict.TryAdd("", "foo");
+
+            // Assert
+            Assert.True(result);
+        }
+
+        [Fact]
+        public void TryAdd_EmptyStorage_CanAdd()
+        {
+            // Arrange
+            var dict = new AdaptiveCapacityDictionary<string, object>();
+
+            // Act
+            var result = dict.TryAdd("key", "value");
+
+            // Assert
+            Assert.True(result);
+            Assert.Collection(
+                dict._arrayStorage,
+                kvp => Assert.Equal(new KeyValuePair<string, object?>("key", "value"), kvp),
+                kvp => Assert.Equal(default, kvp),
+                kvp => Assert.Equal(default, kvp),
+                kvp => Assert.Equal(default, kvp));
+        }
+
+        [Fact]
+        public void TryAdd_ArrayStorage_CanAdd()
+        {
+            // Arrange
+            var dict = new AdaptiveCapacityDictionary<string, object>()
+            {
+                { "key0", "value0" },
+            };
+
+            // Act
+            var result = dict.TryAdd("key1", "value1");
+
+            // Assert
+            Assert.True(result);
+            Assert.Collection(
+                dict._arrayStorage,
+                kvp => Assert.Equal(new KeyValuePair<string, object?>("key0", "value0"), kvp),
+                kvp => Assert.Equal(new KeyValuePair<string, object?>("key1", "value1"), kvp),
+                kvp => Assert.Equal(default, kvp),
+                kvp => Assert.Equal(default, kvp));
+        }
+
+        [Fact]
+        public void TryAdd_ArrayStorage_DoesNotAddWhenKeyIsPresent()
```

Comment:
Add a test that adds 5 items with `TryAdd` and then tests that dictionaryStorage contains them all and arrayStorage is empty.

Reply:
https://github.com/dotnet/aspnetcore/pull/31360/files#diff-57563963409bf2ebaede4548d903e4f87a5415f5530bc7024dec61f94d1979b9R1227

Need to update the test 😄

Reply:
What was the comment pointing to?

I keep mentioning TryAdd because I'm pretty sure it will break when adding an item that causes it to switch from using an array to a dictionary.